enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,940 @@
+
+As already mentioned, backing up and restoring data will happen on system boot.
+It means that next boot makes backup compatible with previously booted
+commit/MicroShift, but it restores data compatible with itself (currently running).
+
+Decision whether to backup or restore will be primarily based on health of previous boot.
+This means that MicroShift needs to keep history of running ostree commits
+(featuring MicroShift) and their health. The software also needs access to the
+history information to know when database format migrations are needed.
+To support both decisions, a structured text file outside of the etcd database
+will be used to persist the history information between various deployments.
+
+Information about system's health will be obtained by greenboot integration
+in form of "green" (system is healthy) and "red" (system is unhealthy) scripts
+which will persist the overall system's health for the current ostree commit.
+
+### Backing up and restoring MicroShift data
+
+MicroShift needs to fit into workflow of ostree-based systems: device can be
+upgraded by staging and booting new ostree commit, and rolled back if it's
+unhealthy or admin wishes to do it.
+
+To achieve that MicroShift needs to make backups and be able to restore them in
+sync with ostree lifecycle.
+As mentioned previously these actions will happen on boot, just before MicroShift cluster runs.
+As a general rule: if previous boot was healthy, data will be backed up, if boot
+was unhealthy, data will be restored.
+In case of manual rollback data should be restored, even if previous boot was
+healthy. The current database will be backed up before being replaced.
+
+To provide ability to rollback and restore MicroShift data, backup for each
+ostree commit/deployment will be kept.
+For now, default behavior will be to keep backups for commits no longer existing on the system.
+Reason for this is that reintroducing commit results in the same ID (it is a checksum)
+and deleting backups would prevent restoring healthy data for the commit.
+This could be a configurable option, but rules of pruning old backups are out
+of scope of this enhancement.
+
+It is worth noting that although enhancement focuses on MicroShift's data,
+backups will be tied to specific ostree commits.
+Linking backups to ostree commits will ensure that staging and rolling back
+is "all or nothing" and MicroShift does not accidentally run applications
+belonging to another commits. Especially that difference between two commits
+might not be MicroShift itself, but the applications that run on top of it.
+
+#### Backup technique
+
+MicroShift will perform backup and restore by leveraging functionality
+Copy-on-Write (CoW). It is a feature of filesystem and is utilized by
+providing a `--reflink=` param to `cp` program.
+Because not all filesystems support CoW, we will provide `auto` argument
+to `--reflink` so it gracefully falls back to regular copy.
```

Comment:
We consider MicroShift to not be very dynamic environment in terms of operations in etcd: we only bring handful of OpenShift's CRDs; there's no MCO to manage host via CRs; user applications wouldn't write to microshift's etcd - they'll use PVCs with LVMS (topolvm) that we ship by default

@dusk125 did some benchmarking of etcd on microshift to test fragmentation and I think he might have some insight on how much data apiserver would need to stuff into etcd to make a difference (although if device works for year or longer, this might be a concern)

@dhellmann correct me if I'm wrong

Reply:
I think @pmtk means they won't access etcd directly, only through the API server.

The trade-off with CoW vs. snapshots came down to several factors. The snapshots are roughly statically sized, need etcd to be running to extract the snapshot, and need etcdctl to restore. The CoW option may vary in size, but does not require any specialized tools. We may change approach later, if the copies prove to be a problem, but we do expect to prune backups over time and that may be more practical than solving the (potential) version-compatibility problems of using specialized tools.

Reply:
i wasn't intending to push back against the use of CoW, only wondering aloud whether there will be any savings from it(reusing shared files that are unchanged), or effectively each backup will require a unique copy of the content.  W/ the result having implications for how much storage the backups take up.

Fundamentally i think i'm just asking "are we making assumptions the storage requirement for backups will be small due to CoW(and the implied sharing/reuse of content between backups), and if so, has that assumption been validated?"

Reply:
I wonder if `du` is confused by the reflinks. I'm seeing 96K difference in space consumed.

```
# df -k .
Filesystem            1K-blocks    Used Available Use% Mounted on
/dev/mapper/rhel-root  51355136 9984136  41371000  20% /

# cp --recursive --preserve --reflink=auto /var/lib/microshift /var/lib/microshift-backups

# df -k .
Filesystem            1K-blocks    Used Available Use% Mounted on
/dev/mapper/rhel-root  51355136 9984232  41370904  20% /

# python -c 'print(41371000 - 41370904)'
96
```

Reply:
We should add something to the risks section about the size of the backups being variable because of CoW and potentially taking up more space on systems with a lot of writes to the API server. Mitigation would be to clean up unused backups more often. We will need to provide instructions for users about how to do that, and may need an `admin prune-backups` command or something.

Reply:
Added